rust/chains/hyperlane-cosmos/src/providers/rpc.rs

```diff
@@ -61,41 +67,101 @@ impl CosmosWasmIndexer {
         locator: ContractLocator,
         event_type: String,
         reorg_period: u32,
-    ) -> Self {
-        Self {
-            conf,
-            _domain: locator.domain.clone(),
-            address: locator.address,
-            event_type,
+    ) -> ChainResult<Self> {
+        let client = HttpClient::builder(conf.get_rpc_url().parse()?)
+            // Consider supporting different compatibility modes.
+            .compat_mode(CompatMode::latest())
+            .build()?;
+        Ok(Self {
+            client,
+            contract_address: locator.address,
+            contract_address_bech32: digest_to_addr(locator.address, conf.get_prefix().as_str())?,
+            target_event_kind: format!("{}-{}", Self::WASM_TYPE, event_type),
             reorg_period,
-        }
+        })
     }
+}
 
-    /// get rpc client url
-    fn get_conn_url(&self) -> ChainResult<String> {
-        Ok(self.conf.get_rpc_url())
+impl CosmosWasmIndexer {
+    #[instrument(level = "trace", err, skip(self))]
+    async fn tx_search(&self, query: Query, page: u32) -> ChainResult<TxSearchResponse> {
+        Ok(self
+            .client
+            .tx_search(query, false, page, PAGINATION_LIMIT, Order::Ascending)
+            .await?)
     }
 
-    /// get contract address
-    pub fn get_contract_addr(&self) -> ChainResult<String> {
-        verify::digest_to_addr(self.address, self.conf.get_prefix().as_str())
+    // Iterate through all txs, filter out failed txs, find target events
+    // in successful txs, and parse them.
+    fn handle_txs<T>(
+        &self,
+        txs: Vec<tx::Response>,
+        parser: for<'a> fn(&'a Vec<EventAttribute>) -> ChainResult<Option<ParsedEvent<T>>>,
+    ) -> ChainResult<impl Iterator<Item = (T, LogMeta)> + '_>
+    where
+        T: 'static,
+    {
+        let logs_iter = txs
+            .into_iter()
+            .filter(|tx| {
+                // Filter out failed txs
+                if tx.tx_result.code.is_err() {
+                    debug!(tx_hash=?tx.hash, "Indexed tx has failed, skipping");
+                    false
+                } else {
+                    true
+                }
+            })
+            .map(move |tx| {
+                // Iter through all events in the tx, looking for the target event.
```

Comment:
Could remove this comment if the closure is moved to its own function named appropriately, and we'd also get less scary nesting